### PR TITLE
controller&service(and also some mistakes)

### PR DESCRIPTION
**Pull request by pustoj0**

### CODE REVIEW

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 1]
```diff
@@ -4,111 +4,202 @@
...
+        this.warehousePremiumLoadRowMapper = warehousePremiumLoadRowMapper;
         this.jdbcTemplate = jdbcTemplate;
     }
     @Override
     public List<WarehouseLoadDto> findWarehouseWidgets() {
         try {
             return jdbcTemplate.query(Queries.SQL_FIND_WAREHOUSE_WIDGETS, warehouseLoadRowMapper);
         }catch (DataAccessException e) {
-            throw crudException("WarehouseWidget",e.toString());
+            throw new CRUDException("DashboardDaoImpl failed to find WarehouseWidget",e);
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: ...to find Warehouse load instead of Warehouse widget.

Widget is a part of the dashboard, so better to call methods by mentioning what we are going to select

File: src/main/java/com/ita/if103java/ims/dto/PopularBodyDto.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package com.ita.if103java.ims.dto;
+
+
+import com.ita.if103java.ims.entity.DateType;
+import com.ita.if103java.ims.entity.PopType;
+
+public class PopularBodyDto {
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dto/PopularBodyDto.java]: please rename to PopularItemsRequestDto

File: src/main/java/com/ita/if103java/ims/dto/PopularityListDto.java [hunk 3]
```diff
@@ -1,6 +1,8 @@
 package com.ita.if103java.ims.dto;
 
+import org.springframework.stereotype.Component;
 
+@Component
 public class PopularityListDto {
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dto/PopularityListDto.java]: Please rename to PopularItemsDto

File: src/main/java/com/ita/if103java/ims/dto/RefillListDto.java [hunk 4]
```diff
@@ -1,18 +1,31 @@
 package com.ita.if103java.ims.dto;
 
+import org.springframework.stereotype.Component;
 
+@Component
 public class RefillListDto {
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dto/RefillListDto.java]: please rename to EndingItemsDto

File: src/main/java/com/ita/if103java/ims/exception/CRUDException.java [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package com.ita.if103java.ims.exception;
 
+
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/exception/CRUDException.java]: not needed change, please remove

File: src/main/java/com/ita/if103java/ims/exception/CRUDException.java [hunk 6]
```diff
@@ -17,4 +18,4 @@ public CRUDException(Throwable cause) {
         super(cause);
     }
 
-}
\ No newline at end of file
+}
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/exception/CRUDException.java]: I think the same as above, please revert this change

File: src/main/java/com/ita/if103java/ims/mapper/jdbc/RefillListRowMapper.java [hunk 7]
```diff
@@ -12,8 +12,10 @@
     @Override
     public RefillListDto mapRow(ResultSet resultSet, int rowNum) throws SQLException {
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/mapper/jdbc/RefillListRowMapper.java]: Please rename this mapper to EndingItemsRowMapper

File: src/main/java/com/ita/if103java/ims/service/DashboardService.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+import com.ita.if103java.ims.dto.RefillListDto;
+import com.ita.if103java.ims.dto.WarehouseLoadDto;
+import com.ita.if103java.ims.entity.DateType;
+import com.ita.if103java.ims.entity.PopType;
+
+import java.util.List;
+
+public interface DashboardService {
+    List<WarehouseLoadDto> getWarehouseWidget();
+    List<PopularityListDto> getPopularityList(int quantity, DateType dateType, PopType popType, String date);
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/service/DashboardService.java]: as I said earlier, better to receive DTO here than 4 separate arguments

File: src/main/java/com/ita/if103java/ims/service/impl/DashboardServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    public List<PopularityListDto> getPopularityList(int quantity, DateType dateType, PopType popType, String date) {
+        return dashboardDao.findPopularityItems(quantity, dateType, popType, date);
+    }
+
+
+    @Override
+    public List<RefillListDto> getRefillList(int minQuantity) {
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/service/impl/DashboardServiceImpl.java]: please rename to getEndingItems

File: src/main/java/com/ita/if103java/ims/service/impl/DashboardServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        this.dashboardDao = dashboardDao;
+    }
+
+    @Override
+    public List<WarehouseLoadDto> getWarehouseWidget() {
+        return dashboardDao.findWarehouseWidgets();
+    }
+
+    @Override
+    public List<PopularityListDto> getPopularityList(int quantity, DateType dateType, PopType popType, String date) {
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/service/impl/DashboardServiceImpl.java]: getPopularItems

File: src/main/java/com/ita/if103java/ims/mapper/jdbc/PopularityListRowMapper.java [hunk 11]
```diff
@@ -1,19 +1,19 @@
...
+import com.ita.if103java.ims.dto.PopularItemsDto;
 import org.springframework.jdbc.core.RowMapper;
 import org.springframework.stereotype.Component;
 
 import java.sql.ResultSet;
 import java.sql.SQLException;
 
 @Component
-public class PopularityListRowMapper implements RowMapper<PopularityListDto> {
+public class PopularityListRowMapper implements RowMapper<PopularItemsDto> {
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/mapper/jdbc/PopularityListRowMapper.java]: PopularItemsRowMapper

File: src/main/java/com/ita/if103java/ims/controller/DashboardController.java [hunk 12]
```diff
@@ -25,8 +26,14 @@ public DashboardController(DashboardService dashboardService) {
     }
 
     @RequestMapping(value = "/warehouseLoad")
-    public List<WarehouseLoadDto> getWarehouseLoad(){
-        return dashboardService.getWarehouseLoad();
+    public List<WarehouseLoadDto> getWarehouseLoad(@RequestParam Long accountId){
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/controller/DashboardController.java]: When Spring security will be added, please use account ID from the UserDetailsImpl class

File: src/main/java/com/ita/if103java/ims/controller/DashboardController.java [hunk 13]
```diff
@@ -25,8 +26,14 @@ public DashboardController(DashboardService dashboardService) {
...
+        return dashboardService.getWarehouseLoad(accountId);
+    }
+
+    @RequestMapping(value="/premiumLoad")
+    public WarehousePremiumStructDto getPreLoad(@RequestParam Long id,
+                                                @RequestParam Long accountId){
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/controller/DashboardController.java]: When Spring security will be added, please use account ID from the UserDetailsImpl class. and remove from request param.

This advise related to all endpoints

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 14]
```diff
@@ -27,32 +28,35 @@
...
-        this.popularityListRowMapper = popularityListRowMapper;
+        this.popularItemsRowMapper = popularItemsRowMapper;
         this.warehouseLoadRowMapper = warehouseLoadRowMapper;
         this.warehousePremiumLoadRowMapper = warehousePremiumLoadRowMapper;
+        this.chargeCapacityRowMapper = chargeCapacityRowMapper;
         this.jdbcTemplate = jdbcTemplate;
     }
     @Override
-    public List<WarehouseLoadDto> findWarehouseLoad() {
+    public List<WarehouseLoadDto> findWarehouseLoadByAccountId(Long accountId) {
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: No neeed to rename method in this case, but it is up to you

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 15]
```diff
@@ -27,32 +28,35 @@
...
         try {
-            return jdbcTemplate.query(Queries.SQL_FIND_WAREHOUSE_LOAD, warehouseLoadRowMapper);
+            return jdbcTemplate.query(Queries.SQL_FIND_WAREHOUSE_LOAD_BY_ACCOUNT_ID, warehouseLoadRowMapper, accountId);
         }catch (DataAccessException e) {
-            throw new CRUDException("DashboardDaoImpl failed to find WarehouseLoad", e);
+            throw crudException(e.toString(),"findWarehouseLoad", "accountId = " + accountId);
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: please revert this changes back, because original exception is lost. Or pass exception object into that method

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 16]
```diff
@@ -162,43 +174,62 @@ public void fillList(WarehousePremiumStructDto wpld, List<WarehousePremiumStruct
...
+            "SELECT wcap.id, ifnull(charge,0) charge, capacity " +
+                "FROM " +
+                "(SELECT wh.top_warehouse_id id, sum(capacity) capacity, account_id, active " +
+                "FROM warehouses wh WHERE account_id=? and active=1 GROUP BY wh.top_warehouse_id) AS wcap " +
+                "LEFT OUTER JOIN " +
+                "(SELECT top_warehouse_id id, sum(quantity*volume) charge " +
+                "FROM saved_items si, items it, warehouses wh " +
+                "WHERE si.item_id=it.id and si.warehouse_id=wh.id " +
+                "GROUP BY top_warehouse_id) AS wcharge " +
+                "ON wcap.id=wcharge.id ";
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: Why did you do so drastically changes to what already worked? Please explain me in ITA

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 17]
```diff
@@ -162,43 +174,62 @@ public void fillList(WarehousePremiumStructDto wpld, List<WarehousePremiumStruct
...
+
+        static final String SQL_FIND_BOT_LOAD=
+            "SELECT ifnull(sum(quantity*volume),0) charge, ifnull(capacity, 0) capacity " +
+                "FROM saved_items si " +
+                "RIGHT JOIN warehouses wh " +
+                "ON si.warehouse_id=wh.id " +
+                "LEFT JOIN items it " +
+                "ON si.item_id=it.id " +
+                "WHERE wh.id=? and it.account_id=?";
+
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: :open_mouth:

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 18]
```diff
@@ -102,50 +102,50 @@ public WarehousePremiumStructDto getPreLoadByAccounId(Long id, Long accountId) {
...
             return wpld;
         } catch (DataAccessException e) {
-            throw crudException(e.toString(),"getPreLoad","accountId=" + accountId);
+            throw crudException(e,"getPreLoad","accountId=" + accountId);
         }
     }
 
-    private List<WarehousePremiumStructDto> findSub(Long id, int level, Long accountId){
+    private List<WarehousePremiumStructDto> findSub(Long id, Long top_warehouse_id, int level, Long accountId){
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: top_warehouse_id -please rename to follow Java code style

findSub - please do not shorten methods and variables, because your code is hard to read and understand

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 19]
```diff
@@ -102,50 +102,50 @@ public WarehousePremiumStructDto getPreLoadByAccounId(Long id, Long accountId) {
...
-                return wpld;
+            }
+            return wpld;
         } catch (DataAccessException e) {
-            throw crudException(e.toString(),"findSub","*");
+            throw crudException(e,"findSub","*");
+        }
+    }
+
+    private void recFill(WarehousePremiumStructDto wp){
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: recFill - please do not shorten methods and variables, because your code is hard to read and understand

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 20]
```diff
@@ -102,50 +102,50 @@ public WarehousePremiumStructDto getPreLoadByAccounId(Long id, Long accountId) {
...
+            wp.setCapacity(wp.getCapacity() + wp.getChilds().get(i).getCapacity());
         }
     }
 
-    private CRUDException crudException(String message, String operation, String attribute) {
-        CRUDException exception = new CRUDException(message);
-        LOGGER.error("CRUDException exception. Operation:({}) Dashboard ({}) exception. Message: {}", operation, attribute, message);
+    private CRUDException crudException(Exception e, String operation, String attribute) {
+        CRUDException exception = new CRUDException(e);
+        LOGGER.error("CRUDException exception. Operation:({}) Dashboard ({}) exception.", operation, attribute, e);
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: Replace this:
```
CRUDException exception = new CRUDException(e);
LOGGER.error("CRUDException exception. Operation:({}) Dashboard ({}) exception.", operation, attribute, e);
```
with this:
```
CRUDException exception = new CRUDException("CRUDException exception. Operation:(" + operation + ") Attribute (" + attribute + ").", e);
```
No need to log here as it will be logged in the GlobalExceptionHandler

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 21]
```diff
@@ -175,16 +175,21 @@ private CRUDException crudException(String message, String operation, String att
...
+                "JOIN warehouses wh " +
+                "ON si.warehouse_id = wh.id " +
+                "GROUP BY warehouse_id) AS  cha " +
+                "RIGHT JOIN " +
+                "(SELECT id, capacity, account_id, top_warehouse_id " +
+                "FROM warehouses " +
+                "WHERE is_bottom=1) AS cap " +
+                "ON  cha.warehouse_id=cap.id AND  cha.account_id=cap.account_id " +
+                "WHERE cap.account_id=? " +
+                "GROUP BY top_warehouse_id";
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: again change) 
I would like to switch subqueries and use LEFT JOIN, as it was before. If it possible, sure.
Please just reply to me and not change

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 22]
```diff
@@ -102,50 +102,50 @@ public WarehousePremiumStructDto getPreLoadByAccounId(Long id, Long accountId) {
...
+        }
+    }
+
+    private void recFill(WarehousePremiumStructDto wp){
+        for(int i=0; i < wp.getChilds().size(); i++){
+            recFill(wp.getChilds().get(i));
+        }
+        for(int i=0; i<wp.getChilds().size(); i++){
+            wp.setCharge(wp.getCharge() + wp.getChilds().get(i).getCharge());
+            wp.setCapacity(wp.getCapacity() + wp.getChilds().get(i).getCapacity());
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: you can use foreach
```
        for(WarehousePremiumStructDto child : wp.getChilds()) {
            recFill(child);
        }
        for(WarehousePremiumStructDto child : wp.getChilds()) {
            wp.setCharge(wp.getCharge() + child.getCharge());
            wp.setCapacity(wp.getCapacity() + child.getCapacity());
       }
```

In this case code is much more cleaner and understandable

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java [hunk 23]
```diff
@@ -102,50 +102,50 @@ public WarehousePremiumStructDto getPreLoadByAccounId(Long id, Long accountId) {
...
+                wpld.get(i).setLevel(level + 1);
+                wpld.get(i).setChilds(findSub(wpld.get(i).getId(), top_warehouse_id, wpld.get(i).getLevel(), accountId));
+                if (wpld.get(i).getChilds().size()==0){
+                    ChargeCapacity chargeCapacity = jdbcTemplate.queryForObject(Queries.SQL_FIND_BOT_CAPACITY_CHARGE,
+                        chargeCapacityRowMapper, top_warehouse_id, wpld.get(i).getId(), accountId);
+                    wpld.get(i).setCharge(chargeCapacity.getCharge());
+                    wpld.get(i).setCapacity(chargeCapacity.getCapacity());
                 }
-                return wpld;
+            }
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: Here you can use simple foreach 
```
            for (WarehousePremiumStructDto warehouseItem : wpld) {
                warehouseItem.setLevel(level + 1);
                warehouseItem.setChilds(findSub(warehouseItem.getId(), top_warehouse_id, warehouseItem.getLevel(), accountId));
                if (warehouseItem.getChilds().size()==0) {
                    ChargeCapacity chargeCapacity = jdbcTemplate.queryForObject(Queries.SQL_FIND_BOT_CAPACITY_CHARGE,
                        chargeCapacityRowMapper, top_warehouse_id, warehouseItem.getId(), accountId);
                    warehouseItem.setCharge(chargeCapacity.getCharge());
                    warehouseItem.setCapacity(chargeCapacity.getCapacity());
                }
            }

```
It will be faster and much more readable

File: src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java (hunk 21 above)
Comment (pustoj0) [reply src/main/java/com/ita/if103java/ims/dao/impl/DashboardDaoImpl.java]: After I delete recursive select, I thought it'll be better to make this one look like other and even if this version is bigger, isn't it more readable
